### ISSUE TITLE
perf: migrate raw tensor allocations to TensorAllocator.Rent and IEngine acceleration

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1014]
## Problem

The codebase has **4,054 raw `new Tensor<T>()` allocations** in hot paths that should use `TensorAllocator.Rent` for pooled memory, and **1,758 scalar `NumOps` loops** in NN layers that should use `IEngine` batch operations.

### Current state
| Area | `new Tensor<T>` | `TensorAllocator.Rent` | `Engine.` calls | Scalar `NumOps` loops |
|------|-----------------|----------------------|-----------------|----------------------|
| NeuralNetworks/Layers | 2,722 | 0 | 5,402 | 1,758 |
| Diffusion | 1,332 | 0 | ~200 | ~100 |
| Clustering | Fixed (ComputeInline) | 0 | minimal | Fixed |

### Impact
- **CURE clustering**: 20 min → 2 sec after eliminating Vector allocations (600x)
- **AgglomerativeClustering**: 30 sec → 2 sec (15x)
- **Diffusion tests**: 15+ sec per test for a [1,4] tensor — dominated by allocation overhead
- Every `new Tensor<T>()` in a forward pass loop creates GC pressure that compounds across 50 denoising steps × N layers

### Phase 1: Forward pass tensor reuse (highest impact)
- NN layers (`DenseLayer`, `ConvolutionalLayer`, `AttentionLayer`, etc.): replace intermediate `new Tensor<T>` with `TensorAllocator.Rent` + `TensorAllocator.Return` for temporaries
- Pre-allocate output tensors at layer construction, resize only when shape changes

### Phase 2: IEngine batch operations
- Replace scalar `NumOps.Add/Multiply/Subtract` loops with `Engine.Add/Multiply/Subtract` for vector/tensor-level SIMD acceleration
- Key targets: activation functions, normalization layers, attention score computation

### Phase 3: TensorPool for training buffers
- Gradient tensors, momentum buffers, optimizer state — use pooled allocation
- Cache backward pass intermediates instead of reallocating each training step

## Files with highest allocation counts (top 10)
Priority files for migration based on `new Tensor<T>` count in forward/backward paths.

## Pattern
```csharp
// BEFORE (allocates per call):
var output = new Tensor<T>(outputShape);

// AFTER (pooled, zero-initialized):
var output = TensorAllocator.Rent<T>(outputShape);
// ... use output ...
// TensorAllocator.Return(output); // only for true temporaries
```

## Related
- PR #33 (AiDotNet.Tensors): zero-initialized tensor memory
- `IDistanceMetric.ComputeInline`: allocation-free distance pattern (completed for clustering)
